File: examples/shop/cart/src/pkg/messaging/nats.ts
// The bus over NATS JetStream: the way an event leaves this process for
// another (ADR cart.0008). One stream for the service, `shop-cart`, over every
// subject under `shop.cart.`; the outbox row's topic is the subject, its uuid
// is the message id the stream deduplicates on, and its metadata rides as
// headers, so a subscriber in any language reads the event's name and the
// trace without opening the payload.
import { AckPolicy, jetstream, jetstreamManager, type ConsumerMessages, type JetStreamClient, type JetStreamManager } from "@nats-io/jetstream";
import { connect, headers, type NatsConnection } from "@nats-io/transport-node";
import { type Bus, deliver, type Handler, type Message } from "./bus.ts";

export const STREAM = "shop-cart";
export const SUBJECTS = "shop.cart.>";
/** How long the stream remembers a message id, in nanoseconds: long enough for any relay retry, short enough to cost nothing. */
const DUPLICATE_WINDOW = 2 * 60 * 60 * 1_000_000_000;
const HEADER_MSG_ID = "Nats-Msg-Id";

interface Open {
  nc: NatsConnection;
  js: JetStreamClient;
}

export class NatsBus implements Bus {
  readonly system = "nats";
  private readonly name: string;
  private readonly opening: Promise<Open>;
  private readonly consuming: ConsumerMessages[] = [];

  /** Connects in the background; `ready` is where a failure to surfaces, and `publish` waits on it. */
  constructor(url: string, name: string) {
    this.name = name;
    this.opening = open(url, name);
    // Observed through ready() and every publish; without a handler here a
    // refused connection would be an unhandled rejection before main got to
    // say what was wrong.
    this.opening.catch(() => undefined);
  }

  async ready(): Promise<void> {
    await this.opening;
  }

  /** At least once, like the relay: a repeat within the window is acknowledged as a duplicate and stored once. */
  async publish(message: Message): Promise<void> {
    const { js } = await this.opening;
    const h = headers();
    for (const [key, value] of Object.entries(message.metadata)) h.set(key, value);
    await js.publish(message.topic, JSON.stringify(message.payload), { msgID: message.uuid, headers: h });
  }

  /** A durable consumer named after the subscriber and the subject, so a service that was down reads what it missed. */
  async subscribe(topic: string, handler: Handler): Promise<void> {
    const { js } = await this.opening;
    const durable = `${this.name}-${topic.replaceAll(".", "-")}`;
    const jsm = await js.jetstreamManager();
    await jsm.consumers.add(STREAM, { durable_name: durable, ack_policy: AckPolicy.Explicit, filter_subject: topic });
    const consumer = await js.consumers.get(STREAM, durable);
    const messages = await consumer.consume();
    this.consuming.push(messages);
    void this.pump(messages, handler);
  }

  private async pump(messages: ConsumerMessages, handler: Handler): Promise<void> {
    for await (const m of messages) {
      const metadata: Record<string, string> = {};
      for (const key of m.headers?.keys() ?? []) {
        if (key !== HEADER_MSG_ID) metadata[key] = m.headers?.get(key) ?? "";
      }
      const message: Message = { uuid: m.headers?.get(HEADER_MSG_ID) ?? String(m.seq), topic: m.subject, payload: m.json(), metadata };
      try {
        await deliver(this.system, message, handler);
        m.ack();
      } catch {
        // Redelivered after the server's backoff; the handler is expected to
        // be idempotent for the same reason the relay's repeats are tolerated.
        m.nak();
      }
    }
  }

  async close(): Promise<void> {
    const opened = await this.opening.catch(() => undefined);
    for (const messages of this.consuming) await messages.close();
    await opened?.nc.drain();
  }
}

async function open(url: string, name: string): Promise<Open> {
  const nc = await connect({ servers: url, name });
  await ensureStream(await jetstreamManager(nc));
  return { nc, js: jetstream(nc) };
}

/** The stream is the publisher's to declare; a second declaration of the same stream is not an error. */
async function ensureStream(jsm: JetStreamManager): Promise<void> {
  await jsm.streams.add({ name: STREAM, subjects: [SUBJECTS], duplicate_window: DUPLICATE_WINDOW }).catch(() => jsm.streams.info(STREAM));
}

File: examples/shop/oms/src/pkg/messaging/nats.rs
//! The bus over NATS JetStream: the way an event leaves this process for
//! another, and arrives from one (ADR oms.0001). One stream per service,
//! named for the first two segments of its subjects - `shop-oms` over
//! `shop.oms.>`, and `shop-cart` over the cart's - so a subscriber declares
//! the publisher's stream by the same rule and whichever side comes up first
//! makes it. The outbox row's topic is the subject, its uuid the message id the
//! stream deduplicates on, and its metadata rides as headers.

use std::collections::BTreeMap;
use std::sync::Mutex;
use std::time::Duration;

use async_nats::jetstream::consumer::pull;
use async_nats::jetstream::{self, stream};
use futures::StreamExt;

use super::{Bus, BusError, Handler, Message, deliver};

/// How long a stream remembers a message id: long enough for any relay retry.
const DUPLICATE_WINDOW: Duration = Duration::from_secs(2 * 60 * 60);

pub struct NatsBus {
    name: String,
    client: async_nats::Client,
    jetstream: jetstream::Context,
    tasks: Mutex<Vec<tokio::task::JoinHandle<()>>>,
}

impl NatsBus {
    pub async fn connect(url: &str, name: &str) -> Result<NatsBus, BusError> {
        let client = async_nats::ConnectOptions::new()
            .name(name)
            .connect(url)
            .await
            .map_err(|e| BusError(e.to_string()))?;
        let jetstream = jetstream::new(client.clone());
        Ok(NatsBus {
            name: name.to_string(),
            client,
            jetstream,
            tasks: Mutex::new(vec![]),
        })
    }

    /// The stream a subject belongs to, declared the same way by both ends.
    async fn stream_for(&self, subject: &str) -> Result<stream::Stream, BusError> {
        let (name, subjects) = stream_of(subject);
        self.jetstream
            .get_or_create_stream(stream::Config {
                name,
                subjects: vec![subjects],
                duplicate_window: DUPLICATE_WINDOW,
                ..Default::default()
            })
            .await
            .map_err(|e| BusError(e.to_string()))
    }
}

/// `shop.cart.basket` → the stream `shop-cart` over `shop.cart.>`.
pub fn stream_of(subject: &str) -> (String, String) {
    let parts: Vec<&str> = subject.split('.').collect();
    let head = if parts.len() >= 2 { parts[..2].to_vec() } else { parts.clone() };
    (head.join("-"), format!("{}.>", head.join(".")))
}

impl Bus for NatsBus {
    fn system(&self) -> &'static str {
        "nats"
    }

    /// At least once, like the relay: a repeat within the window is acknowledged as a duplicate and stored once.
    async fn publish(&self, message: Message) -> Result<(), BusError> {
        self.stream_for(&message.topic).await?;
        let mut headers = async_nats::HeaderMap::new();
        headers.insert(async_nats::header::NATS_MESSAGE_ID, message.uuid.as_str());
        for (key, value) in &message.metadata {
            headers.insert(key.as_str(), value.as_str());
        }
        let payload = serde_json::to_vec(&message.payload).map_err(|e| BusError(e.to_string()))?;
        let ack = self
            .jetstream
            .publish_with_headers(message.topic.clone(), headers, payload.into())
            .await
            .map_err(|e| BusError(e.to_string()))?;
        ack.await.map_err(|e| BusError(e.to_string()))?;
        Ok(())
    }

    /// A durable consumer named after the subscriber and the event, filtered
    /// on the subject, so a service that was down reads what it missed.
    async fn subscribe(&self, topic: &str, event: &str, handler: Handler) -> Result<(), BusError> {
        let wanted = event.to_string();
        let stream = self.stream_for(topic).await?;
        let durable = format!("{}-{}", self.name, event.replace('.', "-"));
        let consumer = stream
            .get_or_create_consumer(
                &durable,
                pull::Config {
                    durable_name: Some(durable.clone()),
                    filter_subject: topic.to_string(),
                    ack_policy: jetstream::consumer::AckPolicy::Explicit,
                    ..Default::default()
                },
            )
            .await
            .map_err(|e| BusError(e.to_string()))?;
        let mut messages = consumer.messages().await.map_err(|e| BusError(e.to_string()))?;
        let system = self.system();
        let task = tokio::spawn(async move {
            while let Some(next) = messages.next().await {
                let Ok(m) = next else { continue };
                let mut metadata = BTreeMap::new();
                let mut uuid = String::new();
                if let Some(headers) = &m.headers {
                    for (key, values) in headers.iter() {
                        let value = values.first().map(|v| v.to_string()).unwrap_or_default();
                        if key.to_string() == async_nats::header::NATS_MESSAGE_ID.to_string() {
                            uuid = value;
                        } else {
                            metadata.insert(key.to_string(), value);
                        }
                    }
                }
                let payload = serde_json::from_slice(&m.payload).unwrap_or(serde_json::Value::Null);
                let message = Message {
                    uuid,
                    topic: m.subject.to_string(),
                    payload,
                    metadata,
                };
                if message.event_name() != wanted {
                    // Another event of the same aggregate: not this
                    // subscriber's, acknowledged unread.
                    let _ = m.ack().await;
                    continue;
                }
                match deliver(system, message, &handler).await {
                    Ok(()) => {
                        let _ = m.ack().await;
                    }
                    Err(err) => {
                        // Redelivered after the server's backoff; the handler
                        // is expected to be idempotent for the same reason
                        // the relay's repeats are tolerated.
                        tracing::warn!(error = %err, subject = %m.subject, "oms: subscriber failed; the message will come again");
                        let _ = m.ack_with(jetstream::AckKind::Nak(None)).await;
                    }
                }
            }
        });
        if let Ok(mut tasks) = self.tasks.lock() {
            tasks.push(task);
        }
        Ok(())
    }

    async fn close(&self) {
        let tasks: Vec<_> = self.tasks.lock().map(|mut t| t.drain(..).collect()).unwrap_or_default();
        for task in tasks {
            task.abort();
        }
        let _ = self.client.flush().await;
    }
}

#[cfg(test)]
mod tests {
    use super::stream_of;

    #[test]
    fn a_subject_names_its_stream_by_its_first_two_segments() {
        assert_eq!(stream_of("shop.cart.basket"), ("shop-cart".to_string(), "shop.cart.>".to_string()));
        assert_eq!(stream_of("shop.oms.order"), ("shop-oms".to_string(), "shop.oms.>".to_string()));
        assert_eq!(
            stream_of("payments.ledger.payment"),
            ("payments-ledger".to_string(), "payments.ledger.>".to_string())
        );
    }
}
